File: sdk/python/examples/09c_hitl_streaming.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Human-in-the-Loop with Streaming — Console Interactive.

Streams agent events in real time via SSE.  When the agent pauses for
human approval, the user is prompted in the console to approve, reject,
or provide feedback — all through the AgentStream object.

Use case: an ops agent that can restart services (safe) and delete data
(dangerous, requires approval).  The operator watches the agent think
in real time and intervenes only for destructive actions.

Requirements:
    - Conductor server with LLM support
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
    - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
"""

from agentspan.agents import Agent, AgentRuntime, EventType, tool
from settings import settings


@tool
def check_service(service_name: str) -> dict:
    """Check the health of a service."""
    return {"service": service_name, "status": "unhealthy", "uptime": "0m"}


@tool
def restart_service(service_name: str) -> dict:
    """Restart a service. Safe operation, no approval needed."""
    return {"service": service_name, "status": "restarted", "new_uptime": "0m"}


@tool(approval_required=True)
def delete_service_data(service_name: str, data_type: str) -> dict:
    """Delete service data. Destructive — requires human approval."""
    return {"service": service_name, "data_type": data_type, "status": "deleted"}


agent = Agent(
    name="ops_agent",
    model=settings.llm_model,
    tools=[check_service, restart_service, delete_service_data],
    instructions=(
        "You are an operations assistant. You can check, restart, and manage services. "
        "If a service is unhealthy, check it first, then restart it. Only suggest "
        "deleting data if explicitly asked."
    ),
)


if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(agent, "The payments service is down. Check it and restart it.")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(agent)
        # CLI alternative:
        # agentspan deploy --package examples.09c_hitl_streaming
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(agent)

        # Interactive streaming alternative:
        # # stream() starts the workflow and returns an AgentStream —
        # # iterable for events, with HITL controls built in.
        # result = runtime.stream(
        #     agent,
        #     "The payments service is down. Check it, restart it, and clear its stale cache data.",
        # )
        # print(f"Workflow started: {result.execution_id}\n")

        # for event in result:
        #     if event.type == EventType.THINKING:
        #         print(f"  [thinking] {event.content}")

        #     elif event.type == EventType.TOOL_CALL:
        #         print(f"  [tool_call] {event.tool_name}({event.args})")

        #     elif event.type == EventType.TOOL_RESULT:
        #         print(f"  [tool_result] {event.tool_name} -> {event.result}")

        #     elif event.type == EventType.WAITING:
        #         print(f"\n--- Approval required ---")
        #         choice = input("  Approve? (y/n/message): ").strip().lower()
        #         if choice == "y":
        #             result.approve()
        #             print("  Approved!\n")
        #         elif choice == "n":
        #             reason = input("  Rejection reason: ").strip()
        #             result.reject(reason or "Rejected by operator")
        #             print("  Rejected.\n")
        #         else:
        #             # Anything else is treated as feedback
        #             result.send(choice)
        #             print("  Feedback sent.\n")

        #     elif event.type == EventType.GUARDRAIL_PASS:
        #         print(f"  [guardrail] {event.guardrail_name} passed")

        #     elif event.type == EventType.GUARDRAIL_FAIL:
        #         print(f"  [guardrail] {event.guardrail_name} FAILED: {event.content}")

        #     elif event.type == EventType.ERROR:
        #         print(f"  [error] {event.content}")

        #     elif event.type == EventType.DONE:
        #         print(f"\n  [done] {event.output}")

        # # After iteration, the full result is available
        # final = result.get_result()
        # print(f"\nTool calls made: {len(final.tool_calls)}")
        # print(f"Status: {final.status}")

File: sdk/typescript/examples/09c-hitl-streaming.ts
/**
 * Human-in-the-Loop with Streaming — Console Interactive.
 *
 * Streams agent events in real time via SSE. When the agent pauses for
 * human approval, the user is prompted in the console to approve, reject,
 * or provide feedback — all through the AgentStream object.
 *
 * Use case: an ops agent that can restart services (safe) and delete data
 * (dangerous, requires approval). The operator watches the agent think
 * in real time and intervenes only for destructive actions.
 *
 * Requirements:
 *   - Conductor server with LLM support
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
 *   - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
 */

import { z } from 'zod';
import { Agent, AgentRuntime, tool } from '../src/index.js';
import { llmModel } from './settings.js';

const checkService = tool(
  async (args: { serviceName: string }) => {
    return { service: args.serviceName, status: 'unhealthy', uptime: '0m' };
  },
  {
    name: 'check_service',
    description: 'Check the health of a service.',
    inputSchema: z.object({
      serviceName: z.string().describe('Name of the service to check'),
    }),
  },
);

const restartService = tool(
  async (args: { serviceName: string }) => {
    return { service: args.serviceName, status: 'restarted', new_uptime: '0m' };
  },
  {
    name: 'restart_service',
    description: 'Restart a service. Safe operation, no approval needed.',
    inputSchema: z.object({
      serviceName: z.string().describe('Name of the service to restart'),
    }),
  },
);

const deleteServiceData = tool(
  async (args: { serviceName: string; dataType: string }) => {
    return {
      service: args.serviceName,
      data_type: args.dataType,
      status: 'deleted',
    };
  },
  {
    name: 'delete_service_data',
    description: 'Delete service data. Destructive — requires human approval.',
    inputSchema: z.object({
      serviceName: z.string().describe('Name of the service'),
      dataType: z.string().describe('Type of data to delete'),
    }),
    approvalRequired: true,
  },
);

export const agent = new Agent({
  name: 'ops_agent',
  model: llmModel,
  tools: [checkService, restartService, deleteServiceData],
  instructions:
    'You are an operations assistant. You can check, restart, and manage services. ' +
    'If a service is unhealthy, check it first, then restart it. Only suggest ' +
    'deleting data if explicitly asked.',
});

// Only run when executed directly (not when imported for discovery)
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(agent, 'The payments service is down. Check it and restart it.');
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(agent);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples --agents ops_agent
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(agent);

    // Interactive streaming alternative:
    // // stream() starts the workflow and returns an AgentStream —
    // // iterable for events, with HITL controls built in.
    // const streamHandle = await runtime.stream(
    // agent,
    // 'The payments service is down. Check it, restart it, and clear its stale cache data.',
    // );
    // console.log(`Execution started: ${streamHandle.executionId}\n`);

    // for await (const event of streamHandle) {
    // switch (event.type) {
    // case 'thinking':
    // console.log(`  [thinking] ${event.content}`);
    // break;

    // case 'tool_call':
    // console.log(
    // `  [tool_call] ${event.toolName}(${JSON.stringify(event.args)})`,
    // );
    // break;

    // case 'tool_result':
    // console.log(
    // `  [tool_result] ${event.toolName} -> ${JSON.stringify(event.result)}`,
    // );
    // break;

    // case 'waiting':
    // console.log(`\n--- Approval required ---`);
    // // Auto-approve since we can't do interactive stdin
    // console.log('  Auto-approving for demo...');
    // await streamHandle.approve();
    // console.log('  Approved!\n');
    // break;

    // case 'guardrail_pass':
    // console.log(`  [guardrail] ${event.guardrailName} passed`);
    // break;

    // case 'guardrail_fail':
    // console.log(
    // `  [guardrail] ${event.guardrailName} FAILED: ${event.content}`,
    // );
    // break;

    // case 'error':
    // console.log(`  [error] ${event.content}`);
    // break;

    // case 'done':
    // console.log(`\n  [done] ${JSON.stringify(event.output)}`);
    // break;
    // }
    // }

    // // After iteration, the full result is available
    // const final = await streamHandle.getResult();
    // console.log(`\nTool calls made: ${final.toolCalls.length}`);
    // console.log(`Status: ${final.status}`);
  } finally {
    await runtime.shutdown();
  }
}

if (process.argv[1]?.endsWith('09c-hitl-streaming.ts') || process.argv[1]?.endsWith('09c-hitl-streaming.js')) {
  main().catch(console.error);
}
